apps/server/src/modules/task/uc/task.uc.ts

```diff
@@ -79,6 +79,54 @@ export class TaskUC {
 		return response;
 	}
 
+	async finishTask(userId: EntityId, taskId: EntityId): Promise<TaskWithStatusVo> {
+		const user = await this.userRepo.findById(userId, true);
+		const task = await this.taskRepo.findById(taskId);
+
+		if (!this.authorizationService.hasTaskPermission(user, task, TaskParentPermission.read)) {
+			throw new UnauthorizedException();
+		}
+
+		task.finishForUser(user);
+		await this.taskRepo.save(task);
+
+		// add status
+		const status = this.authorizationService.hasOneOfTaskDashboardPermissions(
+			user,
+			TaskDashBoardPermission.teacherDashboard
+		)
+			? task.createTeacherStatusForUser(user)
+			: task.createStudentStatusForUser(user);
+
+		const result = new TaskWithStatusVo(task, status);
+
+		return result;
+	}
+
+	async restoreTask(userId: EntityId, taskId: EntityId): Promise<TaskWithStatusVo> {
+		const user = await this.userRepo.findById(userId, true);
+		const task = await this.taskRepo.findById(taskId);
+
+		if (!this.authorizationService.hasTaskPermission(user, task, TaskParentPermission.read)) {
+			throw new UnauthorizedException();
+		}
+
+		task.restoreForUser(user);
+		await this.taskRepo.save(task);
+
+		// add status
+		const status = this.authorizationService.hasOneOfTaskDashboardPermissions(
+			user,
+			TaskDashBoardPermission.teacherDashboard
+		)
+			? task.createTeacherStatusForUser(user)
+			: task.createStudentStatusForUser(user);
+
+		const result = new TaskWithStatusVo(task, status);
+
+		return result;
+	}
+
```

Comment:
Code dupplication!

Reply:
Yes, I know. It's basically the ascpects of entity loading, permission checking and DTO creatiion that have to be generalized.

Reply:
Maybe in fist step create a private method in uc?